CIP-0031/README.md

```diff
@@ -0,0 +1,236 @@
+For example, an oracle provider would need to distinguish the outputs that they create (with good data) from outputs created by adversaries (with bad data). 
+They can do this with a token, so long as scripts can then see the token!
+
+### Hydra
+
+We want reference inputs to be usable inside Hydra heads.
+This raises a worry: since Hydra processes independent parts of the transaction graph in parallel, what happens if it accepts one transaction that references an output, and one that spends the output, but then the first transaction gets put into a checkpoint before the second?
+
+Fortunately, Hydra already has to deal with double-spend conflicts of this kind (although in the naive protocol this results in a decommit).
+This proposal simply introduces a new kind of conflict, a reference-spend conflict.
+Reference-spend conflicts should be handled by the same mechanism that is used to handle double-spend conflicts.
+
+### Controlling referencing
+
+One thing that a user might want to do is to control who can reference an output.
+For example, an oracle provider might want to only allow a transaction to reference a particular output if the transaction also pays them some money.
+
+Reference inputs alone do _not_ provide any way to do this.
+Another mechanism would be required, but there is no consensus on what the design should be, so it is currently out of scope for this proposal.
+A brief summary of a few options and reasons why they are not obvious choices is included below.
+
+A key issue is that the choice to control referencing must lie with the _creator_ of the output, not the _spender_.
+Therefore we _must_ include some kind of change to outputs so that the creator can record their requirements.
+
+#### Check inputs
+
+A "check input" is like a reference input except that the spending conditions _are_ checked.
+That is, it acts as proof that you _could_ spend the input, but does not in fact spend it.
+
+Since check inputs cause validator scripts to be run, it seems like they could allow us to control referencing.
+There are two wrinkles:
+
+- The same script would be used for both referencing and spending, overloading the meaning of the validator script. This is still _usable_, however, since the redeemer could be used to indicate which action is being taken.
+- We would need a flag on outputs to say "this output cannot be referenced, but only checked". Exactly what this should look like is an open question, perhaps it should be generic enough to control all the possible ways in which an output might be used (of which there would be three).
+
+#### Referencing conditions
+
+"Referencing conditions" would mean adding a new field to outputs to indicate under what conditions the output may be referenced.
+This could potentially be an entire additional address, since the conditions might be any of the normal spending conditions (public key or script witnessing).
+
+However, this would make outputs substantially bigger and more complicated.
+
```

Comment:
I think it's cleaner to add an optional "reference validator script" field rather than having "check inputs". How much overhead will an empty field add to the serialisation of a UTXO?

Reply:
I also prefer the optional "reference validator script" field, it does seem cleaner. To be really explicit about the behavior: If the reference validation script was absent it would be the same as validation passing. I like the model of having separate validators for the two semantic actions of referencing vs spending.